### PR TITLE
SILParser: Fix for removal of parse-time name lookup

### PR DESCRIPTION
**Pull request by slavapestov**
The SIL parser parses "dismebodied" `GenericParamList`s that are not associated with a `Decl`, and do not appear in the ASTScope tree. The parser would bind `IdentTypeRepr`s that referenced these generic parameter lists using parse-time lookup.

In order to make this work when parse-time lookup is disabled, we have to plumb `GenericParamList`s through the SIL parser in various places, and also teach Sema's `resolveType()` to handle them.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/Sema/TypeCheckType.cpp [hunk 1]
```diff
@@ -2809,13 +2850,20 @@ Type TypeResolver::resolveSILBoxType(SILBoxTypeRepr *repr,
...
-          getDeclContext(), env, options, resolution.getUnboundTypeOpener());
+          getDeclContext(), genericEnv, options,
+          resolution.getUnboundTypeOpener());
     }
 
-    TypeResolver fieldResolver{fieldResolution};
+    TypeResolver fieldResolver{fieldResolution,
+                               genericParams};
     for (auto &fieldRepr : repr->getFields()) {
+      fieldRepr.getFieldType()->dump();
```
Comment (zoecarver) [inline lib/Sema/TypeCheckType.cpp:2866]: Remnants of a debug session?

Comment (slavapestov) [reply lib/Sema/TypeCheckType.cpp:2866]: Yes, thank you!

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Comment by slavapestov** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**